orbit/pkg/setup_experience/setup_experience.go

```diff
@@ -13,28 +22,227 @@ type Client interface {
+// If the setup experience is supposed to run, it will launch a single swiftDialog instance and then
+// update that instance based on the results from the /orbit/setup_experience/status endpoint.
 type SetupExperiencer struct {
 	OrbitClient Client
+	closeChan   chan struct{}
+	rootDirPath string
+	// Note: this object is not safe for concurrent use. Since the SetupExperiencer is a singleton,
+	// its Run method is called within a WaitGroup,
+	// and no other parts of Orbit need access to this field (or any other parts of the
+	// SetupExperiencer), it's OK to not protect this with a lock.
+	sd      *swiftdialog.SwiftDialog
+	started bool
 }
 
-func NewSetupExperiencer(client Client) *SetupExperiencer {
-	return &SetupExperiencer{OrbitClient: client}
+func NewSetupExperiencer(client Client, rootDirPath string) *SetupExperiencer {
+	return &SetupExperiencer{
+		OrbitClient: client,
+		closeChan:   make(chan struct{}),
+		rootDirPath: rootDirPath,
+	}
 }
 
 func (s *SetupExperiencer) Run(oc *fleet.OrbitConfig) error {
 	// We should only launch swiftDialog if we get the notification from Fleet.
+	_, binaryPath, _ := update.LocalTargetPaths(
+		s.rootDirPath,
+		"swiftDialog",
+		update.SwiftDialogMacOSTarget,
+	)
+
+	if _, err := os.Stat(binaryPath); err != nil {
+		return nil
+	}
+
 	if !oc.Notifications.RunSetupExperience {
 		log.Debug().Msg("skipping setup experience")
 		return nil
 	}
 
-	// poll the status endpoint
-	_, err := s.OrbitClient.GetSetupExperienceStatus()
+	// Poll the status endpoint. This also releases the device if we're done.
+	payload, err := s.OrbitClient.GetSetupExperienceStatus()
 	if err != nil {
 		return err
 	}
 
-	// TODO: fill this in!
+	// If swiftDialog isn't up yet, then launch it
+	if err := s.startSwiftDialog(binaryPath, payload.OrgLogoURL); err != nil {
+		return err
+	}
+
+	// Defer this so that s.started is only false the first time this function runs.
+	defer func() { s.started = true }()
+
+	select {
+	case <-s.closeChan:
+		log.Debug().Str("receiver", "setup_experiencer").Msg("swiftDialog closed")
+		return nil
+	default:
+		// ok
+	}
+
+	// We're rendering the initial loading UI (shown while there are still profiles, bootstrap package,
+	// and account configuration to verify) right off the bat, so we can just no-op if any of those
+	// are not terminal
+
+	if payload.BootstrapPackage != nil && payload.BootstrapPackage.Status == fleet.MDMBootstrapPackagePending {
+		return nil
+	}
+
+	if anyProfilePending(payload.ConfigurationProfiles) {
+		return nil
+	}
+
+	if payload.AccountConfiguration != nil && payload.AccountConfiguration.Status == "pending" {
+		return nil
+	}
+
+	// Now render the UI for the software and script.
+	if len(payload.Software) > 0 || payload.Script != nil {
+		var stepsDone int
+		var prog uint
+		steps := append(payload.Software, payload.Script)
+		for _, r := range steps {
+			item := resultToListItem(r)
+			if s.started {
```

Comment:
instead of doing all or nothing, we should have a set of titles that are in the list, then decide to update them that way instead of if we've passed through the Run method before